### ISSUE TITLE
Linter for grammar

### ISSUE THREAD

**Issue by PierreMarchand20** [tecosaur/LaTeX-Utilities#39]
### Is your feature request related to a problem? Please describe.

Since the issue has been closed in https://github.com/James-Yu/LaTeX-Workshop/issues/881#issuecomment-533659579, I open an issue here in case you are interested in adding more linters.

### Describe the solution you'd like

As I said in the previous link, we could add [TeXtidote](https://github.com/sylvainhalle/textidote) which uses LanguageTools in latex documents. It just strips out latex from your document, calls [LanguageTools](https://languagetool.org) which then checks spell and grammar for a lot of languages. `TeXtidote` then computes where are the locations of the errors/warning in your document. 

There was an issue (solved now) to have an easy-to-parse output (see [here](https://github.com/sylvainhalle/textidote/issues/88)), so I guess it would not be too difficult to use.

It main be expensive to run it at every save for example, but having a command to call it would be nice.

**Comment by tecosaur** [comment]
Could be interesting. Probably more of a medium-term than short-term thing though.

**Comment by tecosaur** [comment]
@PierreMarchand20 something else I'd be interested in in part of speech syntax highlighting. Something like python's textblob may be useful. Would you be interested in working together to add more text/writing tools to this extension?

**Comment by PierreMarchand20** [comment]
I have never worked on a vscode extension and typescript, but that could be interesting.

I looked at Vale but it seems that there are not kind to add support to LaTeX (https://github.com/errata-ai/vale/issues/54).

TextBlob seems interesting, but it only supports three languages (English, and via extensions, French and German) and it seems more focused on sentences analysis.

For the record, LanguageTool works for several languages (https://languagetool.org/languages), and it has also an extension in vscode, but it does not support latex.

In any case, I think we can always redo what Textidote does for calling LanguageTool on latex documents, which is replacing/removing every latex environments using regex and then call any linter we want.

**Comment by tecosaur** [comment]
If you know javascript and are willing to look at docs you'll be good :) if not: no worries.
Long term, I think the best approach would be to implement the latex to plaintext, then translate positions back so that our choice of tools isn't limited. I wonder if we can just ~~steal~~ _find inspiration from_ TeXtidote's implementation.

Then we can leave the installation of individual tools to the end-user, and just supply 'presets' i.e. the TeXtidote thing + output parsing to warnings etc. Perhaps have a setting like this
```json
"latex-utils.externalToolIntegrations.enabled": ["vale", "TextBlob", "LanguageTool"]
```

What are your thoughts?

**Comment by PierreMarchand20** [comment]
I do not know javascript, but I have already done some small contributions to one or two extensions for vscode and it did not seem too complicated ^^ I will try to help.

Yes, I also think it is the best approach! This way we can use any linter with the plaintext output a priori. The only thing we need to do for each linter is to parse their output, which should be not too difficult if it is well correctly done.

**Comment by tecosaur** [comment]
Sounds great! Here's my current plan:
 - Make a new branch for this
 - Lay down the boilerplate
 - Invite you to collaborate
 - We go from there

**Comment by PierreMarchand20** [comment]
Perfect !

**Comment by tecosaur** [comment]
Done! See https://github.com/tecosaur/LaTeX-Utilities/tree/dev-linting

I added vale as an example of how I think it could work.

For input cleaning I think we could start by looking at https://github.com/sylvainhalle/textidote/blob/master/Source/Core/src/ca/uqac/lif/textidote/cleaning/latex/LatexCleaner.java

**Comment by PierreMarchand20** [comment]
Perfect ! I am also quite busy starting a new job but I will also try to take a look ^^ I think it could be really interesting for all the LaTeX writers around here ahah

**Comment by tecosaur** [comment]
I did a bit more with vale to see what sort of things we can hope for, and it's looking good!
![image](https://user-images.githubusercontent.com/20903656/66179112-5102c980-e69a-11e9-9820-3a979964d969.png)
I think the framework is generic enough that we can hope to do something similar with a bunch of tools, once we get the LaTeX to plaintext thing working (and then also a bunch of other issues).

**Comment by PierreMarchand20** [comment]
Ok, I downloaded vale and cloned the repo. I started to play with it a little bit and I have some questions (remember, I have never worked on an vscode extension or with typescript).

- What is the use of the line 
```
vscode.commands.registerCommand('latex-utilities.code-action', (d, r, c, m) =>
            extension.codeActions.runCodeAction(d, r, c, m)
        )
```
- where is it prescribed that the linter should be called when the file is saved ? I noticed that it is used only when I save my document.

**Comment by PierreMarchand20** [comment]
I have not understood what is done in textidote, but here’s how I think we can implement that:
- we need an array of pairs (line number, number of lines) that keeps track of deleted lines.
For example, if we have an align environnement of three lines starting at line 6, the pair would be (6,3).

- we also need a map whose key is the line number and the value is an array of pairs which does the same thing as the previous array, but for characters.

With these two objects, I think we can do the change of coordinates for the diagnostics given by the linter.

I will try to implement that next week.

**Comment by PierreMarchand20** [comment]
But I must say that vale is almost working straightforwardly for my latex documents...

I am not a fan of the output “writing-good” though ^^ I have to try other styles

**Comment by tecosaur** [comment]
Great to see you've started looking at this! I look forward to seeing more of your work.

### Answering your questions

 - Purpose of this line
```ts
vscode.commands.registerCommand('latex-utilities.code-action', (d, r, c, m) =>
            extension.codeActions.runCodeAction(d, r, c, m)
        )
```
Ot simply allows actions to be registered which call (from `codeAction.ts`)
```ts
runCodeAction(document: vscode.TextDocument, range: vscode.Range, source: string, message: string)
```

 - How run-on-save is called

In `main.ts` we have this block
```ts
vscode.workspace.onDidSaveTextDocument((e: vscode.TextDocument) => {
    if (e.uri.fsPath === extension.completionWatcher.snippetFile.user) {
        extension.completionWatcher.loadSnippets(true)
    } else {
        extension.wordCounter.setStatus()
        if (utils.hasTexId(e.languageId)) {
            extension.diagnoser.lintDocument(e) // <--- this line here
        }
    }
}),
```

### Mimicing texidote

This seems like the sort of thing which benefits a lot by chosing a good data structure for it.
I'll have a think about it and let you know if I have any ideas different to yours.

### My own question

How do you find the interface `IDiagnosticSource`? (as implemented in `vale.ts`). If you've got any feedback it would be best to make use of it now rather than later :)

**Comment by tecosaur** [comment]
( btw, I don't think we _should_ need a vale path, because `vale <arguments>` _should_ work, but for some reason it didn't work on my machine )

**Comment by PierreMarchand20** [comment]
Thank you for your answers! Are there any reasons why you need to check `e.uri.fsPath === extension.completionWatcher.snippetFile.user` for calling the linter ?

I will wait for your feedback on the data structure before starting to implement it then ^^

`IDiagnosticSource ` looks nice, the only thing is that I would add somewhere is a way for the user to give its own args to each linter.

**Comment by PierreMarchand20** [comment]
I noticed that Vale has a lot of style to help with styles, but I am not sure it really helps with grammar. LanguageTool also has a java executable standalone [here](https://github.com/languagetool-org/languagetool/tree/master/languagetool-standalone) and it is available in a lot of languages, I suggest we add this one after Vale ^^

**Comment by PierreMarchand20** [comment]
> ( btw, I don't think we _should_ need a vale path, because `vale <arguments>` _should_ work, but for some reason it didn't work on my machine )

I will remove it and let `vale` instead then. I noticed I had to reload vscode the first time to update my PATH variable that must contain the path to 'vale'. Otherwise, try to output 'process.env'

**Comment by PierreMarchand20** [comment]
I started to play around with LanguageTool, the JSON format is defined [here](https://languagetool.org/http-api/swagger-ui/#!/default/post_check), I will try to define the associated linter.

**Comment by tecosaur** [comment]
Let's see
#### What's `e.uri.fsPath === extension.completionWatcher.snippetFile.user` doing?
 - That's just for `extension.completionWatcher.loadSnippets(true)` (live snippet stuff)

#### More command line args
 - Good idea, that's an easy tweak

#### Vale or LanguageTool both have a vscode extension
 - I think they just call command line vale/languagetool within the extension
 - Those are two extensions I've tried/used, the issue I've had is with them + LaTeX.
e.g. `\begin{enumerate}` "enumerate is too wordy", or insude a table `word & word & word` "word is repeated" etc.

#### Data structure

 - Sorry I've forgotten to actually go over that, I'll try to remember to do so tommorow and get back to you :)

#### LanguageTool after Vale
 - Sure :smiley: I just did Vale because I like to build around an example, and I'd done a little bit with vale before (e.g. knew it output json which is nice)
 - Glad to hear you're starting work on that

**Comment by PierreMarchand20** [comment]
Ok for `e.uri.fsPath === extension.completionWatcher.snippetFile.user`, what does it have to do with linters ? Could we put the lines 
```java
if (utils.hasTexId(e.languageId)) {
     extension.diagnoser.lintDocument(e) // <--- this line here
}
```
outside of this check ?

Could you show me where you found the documentation for the vscode diagnostic ? So that I can do the link between the vscode diagnostics and those of LanguageTool.

**Comment by PierreMarchand20** [comment]
I have a first version of LanguageTool.ts, could you look at it ? The only issue was that they do not use position as (line,character), but only give an offset in the whole document. And of course, there are a lot of false positive due to latex :-)

LanguageTool offers a lot of outputs, I just used a small part for now... And I am yet to do the code-actions.

**Comment by PierreMarchand20** [comment]
I just added the code actions, I learned a lot of typescript in a few hours xD

**Comment by PierreMarchand20** [comment]
Last thing and then I stop spamming. The issue is that LanguageTool gives an offset and a length, considering the document as a big string. It is a recurring problem with VSCode so there are functions (`positionAt`) associated with `vscode.TextDocument` to go from this representation to using a pair of (line,offset). Problem: we need to call LanguageTool on `vscode.TextDocument` after we remove all the latex, so I am not sure we will still be able to use `positionAt`.

**Comment by tecosaur** [comment]
Great to see the work you've done, thanks for helping out with this!

#### Tweak if statement in main.ts
 - Done in a0008f51033793bcfe768cc661f15ab463f6641f
#### Documentation for vscode diagnostic
 - here: https://code.visualstudio.com/api/references/vscode-api#Diagnostic
#### Take extra arguments
 - Done in 258c0f4dfa38b7db5f1ef28207f0ebd33e1bc446
#### Ofset + length representation
 - Shouldn't be too hard to work around
 - Also, I was thinking the best way may be to create a temporary file, e.g. `document.tex` → `.document.txt`
#### Testing languagetool
 - Tried it, got `NodeError: stdout maxBuffer length exceeded` first
 - Tried it with a short document, worked!
![image](https://user-images.githubusercontent.com/20903656/66798322-305f2d00-ef40-11e9-8a1a-cce1dd67a0ad.png)
 - May want to do a bit of tinkering with the error serverity
#### New idea
 - I noticed that languagetool may take some time to run, with this in mind I think we might benefit from
   - Making lintDocument async (done in 14859161be12f7d3cd1b9e2e4882a3d702500a3b)
   - Adding a `currentlyRunning` bool attribute so that multiple instances of a long(ish) running command aren't spawned

**Comment by tecosaur** [comment]
Furthermore to the `NodeError: stdout maxBuffer length exceeded` error, I think the way to solve that is by switching from using `execFile` to `spawn`.

**Comment by PierreMarchand20** [comment]
Great!

I think I also got `NodeError: stdout maxBuffer length exceeded`, I guess this is because there are too many false positive with the latex environments. Once we will have solved the issue of removing all the latex, it should not appear again.

What async does ? It seems that it should be used with await usually. I agree the zcurrentlyRunning` idea!

**Comment by tecosaur** [comment]
> I guess this is because there are too many false positive with the latex environments

This is just an issue with long outputs. Using `spawn` we should be able to handle much larger outputs. I'll take a look at that.
> What async does?

Afaik this should just help with it being non-blocking
> I agree the `currentlyRunning` idea!

Shouldn't be too hard to implement :smiley:

**Comment by tecosaur** [comment]
Switched to spawn and implemented `currentlyRunning` (as `currentProcess`) in 75143f05fd18c35a666f88a949529deb6b68a3f4

**Comment by tecosaur** [comment]
Minor update: I tested my big-ish tex file and it all worked.
Also had a think about the data structure and arrays of removed content (one for lines, on for characters) seems like a relatively decent approach to me.

I'm getting a really good feeling about this!

**Comment by tecosaur** [comment]
Hmmm. Another thought, since these tools run on the whole file (and can take a few seconds), it could be nice to make use of a stateful model. I.e. change a few words in a paragraph, just run tools on modified paragraph instead of the full document. Perhaps we could just do this in-between saves with a timeout :thinking:

**Comment by PierreMarchand20** [comment]
Perfect! 

If we can create another file to parse with the linter, it should be not too difficult. But I was afraid it would be too expensive. And where could we create this file ?

I will try what I suggested then ^^

**Comment by PierreMarchand20** [comment]
> Hmmm. Another thought, since these tools run on the whole file (and can take a few seconds), it could be nice to make use of a stateful model. I.e. change a few words in a paragraph, just run tools on modified paragraph instead of the full document. Perhaps we could just do this in-between saves with a timeout 🤔

or we can define a command so that the user decides when he wants to call the linters?

**Comment by PierreMarchand20** [comment]
It would be really interesting to run it only on modified parts, as you said, it takes time. But may be it would be difficult to do it in a first time directly.

**Comment by tecosaur** [comment]
The best bet is probably just to create a file in the tmp folder. See the use of `tmpdir` in `tikzpreview.ts`.

> Too expensive
I think the contrary. We can easily run multiple tools on the same file, and maybe even make incremental changes

> define a command so that the user decides when he wants to call the linters
> It would be really interesting to run it only on modified parts, as you said, it takes time. But may be it would be difficult to do it in a first time directly.

I think first time we just run on the whole file, there's not really much else you can do. Then as part of the linter-config we could have a setting like
```json
"latex-utils.diagnoser.tools": {
  "languagetool": {
    "enabled": true,
    "args": [],
    "run": "onSave" | "onSaveAndIncremental" | "incrementally"
  }
}
```
or similar. Basicly I see there being three main modes:
 1. Lint entire file on save
 2. Lint entire file on save, and incrementally with timeout
 3. After initial whole-file run, run incrementally on save & timeout

**Comment by PierreMarchand20** [comment]
Ok, I will look at `tmpdir` then. 

What do you mean by incremental changes ? Anyway, sounds great to me

**Comment by tecosaur** [comment]
By incremental changes I mean
```diff
- this is is a test
+ this is a test
```
then just rerunning the linters on the changed line.

**Comment by PierreMarchand20** [comment]
Ok I see.

What I meant is that may be we should first implement the linter on the whole file, removing the latex and so on. And then, when it will work, we can try to implement the incremental feature, what do you think ?

**Comment by tecosaur** [comment]
That's what I was thinking :smiley: I just think it's good to keep these things in mind when implementing to we don't unintentionally make our lives harder later on.

On that note, with the LaTeX to plaintext conversion, I think I'd be good to use a very modular/versatile approach for the parsing. I'd like to do some things such as
 - Ignore arguments of specific commands — e.g. `\*ref{...}`
 - Ignore some environments, and their optional arguments  — e.g. `tikzpicture`, `minted`

Given the existence of user-definable commands, I think it would be best to operate on a blacklist model. So I'd imagine the config would be split into two sections, commands and environments. Then for each item have a few properties that set ignoring/passthrough of some/all mandatory/optional arguments This would ideally be able to mix in / be replaced with a user config setting, any single JSON object should work.

From there the best thing to do would probably be just to compile a giant regex expression and run that against the document.

Those are my ideas, what do you think?

**Comment by PierreMarchand20** [comment]
> That's what I was thinking 😃 I just think it's good to keep these things in mind when implementing to we don't unintentionally make our lives harder later on.

I agree perfect!

> On that note, with the LaTeX to plaintext conversion, I think I'd be good to use a very modular/versatile approach for the parsing. I'd like to do some things such as

> Ignore arguments of specific commands — e.g. \*ref{...}
> Ignore some environments, and their optional arguments — e.g. tikzpicture, minted
> Given the existence of user-definable commands, I think it would be best to operate on a blacklist model. So I'd imagine the config would be split into two sections, commands and environments. Then for each item have a few properties that set ignoring/passthrough of some/all mandatory/optional arguments This would ideally be able to mix in / be replaced with a user config setting, any single JSON object should work.

> From there the best thing to do would probably be just to compile a giant regex expression and run that against the document.

I am not sure I understood everything. 

- Concerning commands, why do you want to only ignore arguments?

- What are the properties you want for each item ?

In my mind I was just thinking of a list of commands and a list of environments as you say in a JSON files (so that a user can add his own). And then, for each command 'foo' we remove '\foo[...]{...}' and for each environment 'bar', we remove `\begin{bar}...\end{bar}`.

**Comment by PierreMarchand20** [comment]
Then, there are some details of course (`document` environment, `itemize`/`enumerate` environment that we can keep, `$...$`, etc)

**Comment by tecosaur** [comment]
I feel as though I'm overthinking this I think (for commands) an array of regex is probably best.

Unfortunately due to nesting parenthesis (e.g. `\graphicspath{{...}}`) we can't use just regex :(

To clarify on my terminology, I think of latex macros like this:
 - `\command[optional1][optional2]{mandtory1}{mandatory2}`

**Comment by tecosaur** [comment]
> why do you want to only ignore arguments?

By that I mean that we want to ignore the `\command` bit always, but sometimes we want to pass through an argument (e.g. `\emph{...}`), sometimes we want to ignore it (e.g. `\tikzset{...}`), then a similar thing with environments of course.

Not sure what the cleanest way of being able to represent differences is what we want done would be.

**Comment by PierreMarchand20** [comment]
That is when we can look at what he has done in textidote ^^

**Comment by PierreMarchand20** [comment]
Then, we can define a list of commands where we delete everything, a list of command where we keep the argument. And it is a similar for environments, but it is a little bit more difficult to keep the argument of an environment.

**Comment by tecosaur** [comment]
Perhaps then we just take the approach where, this may be a little heavy-handed, but it will be good for 99% of the document. Remove all optional arguments of environments, and then split commands into "keep arguments" and "remove arguments". That sounds reasonable to me.

**Comment by tecosaur** [comment]
@PierreMarchand20 how's it going?

**Comment by PierreMarchand20** [comment]
I did not have the time to do it so far, may be this weekend.

**Comment by PierreMarchand20** [comment]
Ok I looked a little bit, I think actually this easiest would be to get the text of the file in a string and call regex functions. I do not see the advantage of copy paste the text in an auxiliary file. What do you think ? We can still do every operations on a string and the create an auxiliary file to store the cleaned string if it is really necessary.

With this in mind I wanted to use the function `matchall` (https://javascript.info/regexp-methods) which seems really perfect for us, but I cannot manage to use it. I added the necessary lib as mentioned [here](https://stackoverflow.com/a/57298833/5913047) but still, when I compile, it does find the function. Any ideas ?

**Comment by PierreMarchand20** [comment]
It would be really nice because with `matchall` we get the positions of every matches, their associated capture groups, etc.

**Comment by tecosaur** [comment]
Just two things

#### Just using regex
I'm not sure that will be able to process balanced parenthesis etc. using regex

#### Files
Can we run all the tools we may want to (so far just vale and languagetool) via stdin? If not I think we need a file

**Comment by PierreMarchand20** [comment]
Why could we not use regex for balanced parenthesis ? I already used it to replace all `$...$` by `\(...\)` in my Tex documents with the search and replace of vscode using regex and capture groups. And it may be more efficient to do it directly on a string than processing a file line by line for example. I do not know in typescript but usually reading files can be a bottleneck somehow.

Then, we can still pu the processed string in a text file and then call the linters. But I know there is an example of vale without file for example.

**Comment by tecosaur** [comment]
For an idea of the issues with regex look at this (look at the unit tests)
https://regex101.com/r/dHcL4U/1

For the file thing, how about we just see how far we can go with stdin and then switch to files as needed :)

**Comment by PierreMarchand20** [comment]
I guess we can find every first level, and check if there is some recursive pattern in it. In any case, it is not easy, and even with a file, I think we would end up doing pretty much the same thing.

And do you have any ideas on how to make `matchAll` work ? I think we should be able to use it because it is part of es2020 that seems to be a valid argument for the compiler but I cannot manage to make it work :-/

**Comment by tecosaur** [comment]
`es2020` may be a bit ambitious. I think we can only go up to `es2018` using the vscode's current version of electron. 

AFAIK matching N deep nested parenthesis is _not_ possible using the nodejs regex engine. Others like `PCRE` have things like recursion tokens (`?R`), but javascript doesn't.

**Comment by PierreMarchand20** [comment]
Yes I saw for the recursive regex, but as I said, we can just get the first matches and check if there are other embedded commands in it with a while loop or something.


But if you have a better idea, we can do differently.

**Comment by PierreMarchand20** [comment]
`es2020` should work in vscode, at least it is an available option according to https://code.visualstudio.com/docs/languages/jsconfig

**Comment by PierreMarchand20** [comment]
I just pushed a first bought implementation:

## Main changes
 
- in `latexToPlaintext`  I compute all the changes I want to make in the file in an array `[number,number,number,number]` called `offsets`corresponding to
  - the line where the change has been made
  - the number of deleted lines made by the change
  - the position of the first character modified
  - the length of the modification
  so usually, either the first is different from zero, or the others
- then I copy the modified text in a temporary file (I tried to do what you did for the tiki visualization)
- in `vale.ts` and `LanguageTool.ts`, I call the linters on the temporary file and I use `offsets` to change the offsets and lines. It could be refactored in one function that takes two `Positions`, but I do not know how to define a function for `IDiagnosticSource`.

## Remarks

- in `latexToPlaintext`, I have a list of regex rules that I use. So far, it corresponds to what I want to remove in the file. We could also read the regex rules from a `.json` file (I did not manage to do it). Of course, it needs to be improved (checking recursive patterns in matches, differentiate between removing everything, keeping arguments) and it can be modified if we do not want to use regex, we just need to compute `offsets` and it should work.
- I tried it on parts of some of my articles, and it works well.

**Comment by PierreMarchand20** [comment]
<img width="340" alt="image" src="https://user-images.githubusercontent.com/15056323/67163089-b4f7e400-f362-11e9-9911-7915579e6164.png">

I kept warning for vale and error for LanguageTool to differentiate them for debugging purposes. But we should change that at some point of course.

**Comment by tecosaur** [comment]
Fantastic! I'm afraid I'm a little busy over the next few days but I can't wait to check it out.
Thanks again for being willing to help with this :smiley:

**Comment by PierreMarchand20** [comment]
No worries ! I will wait for your feedback before going further then ^^

**Comment by tecosaur** [comment]
Feel free to keep going — it looks promising! Just don't expect anything more than a cursory glance from me until later this week :stuck_out_tongue:

**Comment by tecosaur** [comment]
Some thoughts from me:

### The config

```json
                "latex-utilities.linter.vale": {
                    "type": "array",
                    "default": [],
                    "markdownDescription": "Arguments for vale."
                },
                "latex-utilities.linter.LanguageTool": {
                    "type": "array",
                    "default": [],
                    "markdownDescription": "Arguments for LanguageTool."
                },
```

To me this is overly specific. I think a more versitile config format would be something like this:
```json
{
  "toolName": ["arg1", "arg2"],
  "anotherTool": ["arg3"]
}
```
I changed this in 64a5819cb9f60149be772496e6be6b8a688b33bb

### Linting

The actual linting didn't work for me. I think your regex took out everything between `\begin{document}` and `\end{document}`

**Comment by PierreMarchand20** [comment]
Sure, I was not sure what kind of types we could have.

And yes, for now it removes everything between `\begin ... \end` environment. As it is said in the implementation, it is a first rough implementation ahah

**Comment by tecosaur** [comment]
A good start :)

**Comment by PierreMarchand20** [comment]
I am pretty sure there is something better to do than a array of size four to describe a change, it is very C++, I guess there is cleaner and more javascript/typescript way of presenting it ?

Again, I am not used to javascript and typescript :-/

**Comment by tecosaur** [comment]
I reckon you could probably use the type `vscode.Range` which is an ordered pair of two `vscode.Position`s, i.e.
```js
new vscode.Range(
    new vscode.Position(lineNo, characterNo),
    new vscode.Position(lineNo, characterNo)
)
```
The slightly nice thing about this is that there are a few nice prototypes such as
 - `isAfterOrEqual(other: Position): boolean;`
 - `compareTo(other: Position): number;`
 - `translate(lineDelta?: number, characterDelta?: number): Position;`

**Comment by PierreMarchand20** [comment]
I am not sure, `offsets` is a little bit trickier than a range. I was thinking of may be we could define a class of our own instead of using these array of size four to characterize a modification of the original TeX file. But may be it is an overkill

**Comment by tecosaur** [comment]
Hmm. Perhaps switching from offsets to ranges could also work?
We could extend the class to provide removed lines etc. more easily.

**Comment by PierreMarchand20** [comment]
Yes this could work. The issue of what I have done so far is that I make the hypothesis that the modification is on one line (one offset and a length) or just several deleted line. May be switching to range would make it more general

**Comment by tecosaur** [comment]
Mmmm. We could just add all the deleted ranged to an array. I think that will be an improvement overall.

**Comment by PierreMarchand20** [comment]
Ok, I have started to modify the code accordingly then, I will push soon

**Comment by PierreMarchand20** [comment]
I pushed a new version using `changes` instead of `offsets`, it is an array of `vscode.Range` as you suggested. 

Like before, the lines 117-136 in `vale.ts` and 64-84 in `language tool.ts` can be factored in a function taking two positions (the one of the issue), and returns the translated two positions to use in the original file. I let you do that because I do not really know where to put this function.

Apart from that, I think the part "computing the right positions" from the issues detected by the linters is done. What needs to be improved now is the detection of all the latex environments we want to get rid off to apply the linters, so the function `latexToPlaintext` in `diagnoser.ts` in other words. 

There are some small issues also, for example  I replace all the latex environments by a dummy character, but it creates some warnings from the linters of "repeated words" sometimes...

**Comment by tecosaur** [comment]
Marvellous! I really appreciate the work you're doing for this. I'd love to have a look, but the next ~two weeks are end of semester exams for me, so unfortunately for the near future I'll have to relegate this to "procrastination material" :stuck_out_tongue: — so basicly I'm still good for small questions, but shouldn't look at much more.

I'll let you know when I'm able to take a look at this, until then feel free to forge ahead :smiley:

**Comment by PierreMarchand20** [comment]
ahahah fair enough, I will see what I can do next then.

**Comment by PierreMarchand20** [comment]
I checked and it seems we cannot define/implement a function in an interface in typescript, so I do not know how to factor the implementation of "applying change", may be changing the interface to an abstract class ?

**Comment by tecosaur** [comment]
I'd just use a class if you need a function prototype, what's the full context for this?

**Comment by PierreMarchand20** [comment]
What do you mean by context ? It is just that the function that computes the coordinates in the original document using the changes we made (the array of vscode.Range) is the same for any linters.

**Comment by tecosaur** [comment]
Ah, so you're talking about `translatePlaintextPosition(linter: IDiagnosticSource)`? (this is a side-effect of me trying to not get dragged into the source code until after exams). Anyway, if you want a type to have prototypes that are functions, just use a class :)
#### Example
```ts
class offset extends vscode.Range {
    public lines() {
        return this.end.line - this.start.line
    }
}
```

#### p.s.
FYI: Interfaces are just for type checking, they can't actually _do_ anything.

**Comment by tecosaur** [comment]
@PierreMarchand20 exams are over! I can now get back to this. How have you been?

**Comment by PierreMarchand20** [comment]
Hey! I hope everything went fine then ! I have to say that I have not done much since my last message, I was pretty busy at work.

But what is left to do is to factorize what you call `translatePlaintextPosition(linter: IDiagnosticSource)`, but I removed it because I thought it was easier to do it directly in `vale` and `LanguageTool`. It allows to directly create`vscode.Range` with the right positions, but may be it is possible to modify the `vscode.Range` afterward.

Then, I think the part "applying change of coordinates" would be done and then we need to improve how to look for LaTeX environments.

I should be able to get back to it this weekend, if you need help.

**Comment by LeCyberDucky** [comment]
Hey, 

I just stumbled upon this by chance (I'm pretty amazed at that coincidence, hah). I'm in the middle of writing a big report in LaTeX, and I have my BSc project coming up pretty soon as well. Therefore, I installed VS Code a few days ago, in order to get a nice setup up and running for writing LaTeX documents. 

With that out of the way, I'll get to my main point: The main thing that my setup is currently missing is some nice grammar and spell checking. I have been thinking about implementing my own solution for this, but now that I have seen what you guys are up to, I think I'll be taking a close look at this first. 

I have a few questions, if you don't mind: 

I see that you have been talking about how checking the file too often might be too resource-heavy. Is that because you are checking the whole file every time? 

The idea for my project was to mimic the behavior of MS Word by checking spelling and grammar in real-time (so, every time a word has been typed and followed by a space). For this, I wanted to check the full file once on startup. After that, I would just continuously check the paragraph (or maybe the line) that is currently being worked on. 

I'm very fond of the idea of creating a nice interface for letting the user supply their own engine for linting. That would fit nicely with what I had in mind, as I wanted to parse the text (that's the tough part for me) and then let Word do the grammar and spell checking for me. 

Also, have you thought about letting the user right-click a highlighed piece of text and then letting them choose a suggested fix? 

I'll try to follow this project (I would love to contribute, if I see something that I can do), and I hope to see some nice progress being made. It looks promising, and I feel like a good implementation of something like this is dearly needed. =)

**Comment by PierreMarchand20** [comment]
So first, I think that even when this project will work, I will still use [Code Spell checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) which is really nice for spell check and works with LaTeX (I finished to write my PhD thesis a few weeks ago :-)). For me the advantage of using any linter we want is to get checks on grammar and style.

For grammar checks for now, we have LanguageTool which takes sometime to process one `.tex`. But for me it makes sense that grammar checks are more expensive than just spell checks. But may be we can find a faster linter for that. At least for now, it runs the linter on save, we could always add an option to do it regularly.

For the right-click, we have done it using vscode actions, so you can do what you want using `Quick fix`, but the shortcut is not right click
<img width="274" alt="image" src="https://user-images.githubusercontent.com/15056323/68896736-0a39d200-0724-11ea-9898-5a12cc6b032f.png">

**Comment by tecosaur** [comment]
> I'll try to follow this project (I would love to contribute, if I see something that I can do), and I hope to see some nice progress being made.

@LeCyberDucky any help is much appreciated! I'm planning on working on this, this weekend, possibly with Pierre's help. If you also would like to give a helping hand, or even just go over the code we've written and made sure everything looks good that would be handy :)

I'd like to try an 'incremental' processing model after one sweep of the whole document, but we'll just have to wait and see how that goes.

Other than that hopefully you'll see this soon-ish in a `v0.4.0`.

**Comment by PierreMarchand20** [comment]
So I am looking back at the code, and the reason why I did not used `translatePlaintextPosition` is that we need to modify the ranges in `linter.actions`, but also in `linter.diagnostics`. It just seemed easier to do it once before creating those two in the Implementation of the linters. 

What do you want to do then ? do we still use `translatePlaintextPosition ` or is there another way to factorise the duplicated lines in `vale` and `LanguageTool` ?

**Comment by PierreMarchand20** [comment]
I just made several updates:

- now I use list of expressions to build the regex used to parse. So that, we just need to add expressions to these lists (and we could add the possibility to the use to add expressions too). There is a list of begin/end environments to remove, a list of commands to remove with their arguments.

- `changes` is now an array of `[vscode.Range, number]`, where the second item is the size of the replacement of the first item.

- Using the last two points, I added a list of commands to replace `\foo{bar}` by `bar`. Applying the translation is a bit tacky, but it works for now. I check that the position of the first character of the change is equal the position of the first character of the output of the linter. For example: in original file
```
I like \foo{aples}
```
in modified copy
```
I like aples
```
If a linter wants to lint `aples`, then the first character of the output is 8 and the first character of the change is also 8.

- I took care of the preamble and some special environments: `$...$` and `\(...\)`

**Comment by tecosaur** [comment]
Marvellous! Unfortunately, my weekend hasn't turned out to be as free as I was hoping for :(

I do however have time to thank you for the improvements you've pushed, and respond to the points you've raised.
 - Thanks :) :) :)
 - Seems like a decent approach. I've just made the begin/end/document regex more robust.
 - Seems reasonable at a first glance
 - I think I may tweak the latex parsing, but the general idea of what you've done looks good :)

Re: `translatePlaintextPosition` and duplicate lines in `value` and `LanguageTool`. One approach that comes to mind is to factor out common/duplicated code into a `generic.ts` component which we can import from.

**Comment by PierreMarchand20** [comment]
No worries! Thank you for letting me work in your extension ^^ I think it will be a nice improvement for LaTeX & VScode users when it will be robust enough.

I think the general structure is modular enough so that you can tweak as much as you want without any issue ^^ I think I will let you factor out common/duplicated code so that it is better done ahah and then we could think to add other linters. 

I agree that the latex parsing still needs improvements, what are your ideas about that ?

There are still some minor issues:

- the environments that are completely removed are replaced by `X`, but it adds some warnings about duplicate words... (when two removed environments are in the same sentence for exemple). 

- one false positive I have is when I finish a sentence with a begin/end environment (`align` or `equation`) for example. Then the final point of the sentence that is in the environment gets removed so that I get a warning about adding a final point... no idea how to fix this one. May be should look in textidote how he does it.

**Comment by tecosaur** [comment]
Sorry for the silence, I've now got a bit more time to spend on this so you should see some commits from me shortly :)

**Comment by tecosaur** [comment]
@PierreMarchand20 I hope you don't mind, but it looks like I'm overhauling your `latexToPlaintext` code.

**Comment by tecosaur** [comment]
Right. It's not _quite_ working but I think my revamp is in the right direction. If you could test it out and tell me what you think that would be great!

**Comment by tecosaur** [comment]
BTW the test document which I'm using (and which seems to be a good test, as I see that there are some issues with the current code) is https://github.com/tecosaur/BMC/blob/master/documentation/bmc.tex

**Comment by PierreMarchand20** [comment]
Ok great I will try on my articles also. But could you explain a little bit what you did ?

May be we could factorize the function to translate the positions in utils.ts now also ?

**Comment by tecosaur** [comment]
I hope the code is fairly easy to read, but then I wrote it :sweat_smile:. I'l try to explain a bit
 - Everything we're interested in starts with `\command`, so we start by matching all of those
 - We loop through all regex matches
 - For each command found we want to identify the arguments
   - We look at the next character after the command, if it's `[` or `{` we find the _balanced_ (this is why I'm not using regex) pair
    - Repeat until all arguments found
 - Run `replaceCommand` or `processEnv`
   - `replaceCommand`
     - removes `\command` and any space after it
     - `transparencyLevel` determined
     - `replaceArgs` called
   - `processEnv`
     - Finds end of env
     - Removes `\begin{env}`
     - Process rest of args using the env's `transparencyLevel`
     - If env is "opaque" remove it's content
 - Here whenever I've said "Remove" or "Replace" I mean calling `queueReplacement`
   - `queueReplacement` just replaces start-end later
   - Makes things easier because you don't need to account for changing positions
   - At the end (:245) all replacements are done in bulk

Let me know if that does the trick.

> Maybe we could factorize the function to translate the positions in utils.ts now also?

Hmm. I'm not sure. The purpose of `utils.ts` is to have functions that get used all over the extension.
While I can see this happening with `getClosingBracket` I don't quite see the same with `translatePlaintextPosition`.

**Comment by tecosaur** [comment]
@PierreMarchand20 any thoughts?

**Comment by PierreMarchand20** [comment]
So I just tried on one of my documents and it does not really work. So I have to go through your changes yet, but apparently you did not change `vale.ts` and `LanguageTool.ts` so I guess the parser do not give the right coordinates or something.

**Comment by PierreMarchand20** [comment]
So the transparency level, it means if we keep the arguments or not right ? I do not feel like the name is right. And do you have an example where it would be useful to keep the optional arguments ?

**Comment by PierreMarchand20** [comment]
There is a lot to take in. I understand the improvement of having your approach is to be able to match matching bracket or parenthesis. It is doable using regex I think (discarding the first answer, several approaches are presented [there](https://stackoverflow.com/questions/546433/regular-expression-to-match-balanced-parentheses)), but why not, may be it would make things easier to deal with recursive calls. But may be we could do the change iteratively ? Like taking a test case that must not be broken ? Because a lot of things look broken know.

A few remarks/questions I have:
- the transparency level, it means if we keep the arguments or not right ? I do not feel like the name is right. 
- do you have an example where it would be useful to keep the optional arguments ?
- it seems that some replacements are done without being added to `this.changes`, which could break everything. For example, lines 197 and 198, for removing '\begin{document}' and '\end{document}'
- You replace usually by nothing, which could lead grammar spellings to errors. In TeXidote, he is usually replacing by `X`, which is what I did. For example
<img width="699" alt="image" src="https://user-images.githubusercontent.com/15056323/70358066-664dcd00-1870-11ea-9127-56c1fe9252f5.png">

- `replacement` seems very similar to `this.change`, is it only for debugging ?

My point is that we had an "almost" working prototype and may be we should work iteratively from it, and not change everything in one go.

**Comment by PierreMarchand20** [comment]
We could store a test file, its cleaned version and the translated positions for example.

**Comment by PierreMarchand20** [comment]
There are issues with the given positions as input, so that the warning are not well translated:
<img width="320" alt="image" src="https://user-images.githubusercontent.com/15056323/70359332-9054be80-1873-11ea-9f81-df1a834f63aa.png">
Since it worked before, I suspect the input. But also in the replace action, the cleaned version has some issues:
<img width="541" alt="image" src="https://user-images.githubusercontent.com/15056323/70359372-a9f60600-1873-11ea-9b26-074152c1f4b2.png">

**Comment by PierreMarchand20** [comment]
Are you sure you pushed everything ? I tried your example and it also does not work.

**Comment by tecosaur** [comment]
Ok, this is going to be a long one (apologies in advance)

> So I just tried on one of my documents and it does not really work.

I thought there were fewer issues. I was wrong. I'm about to push a small commit which should make it a lot better (6bc90aef0e9b695c329abc6476b66b0c6c80e4e9).

> The transparency level, it means if we keep the arguments or not right? I do not feel like the name is right.

Yep that's the idea, with `0` being keep nothing, `3` being bass through all the arguments, and `1` and `2` be pass through only `[]` or `{}` arguments. If you have a better name, then I'm happy to rename it,

> do you have an example where it would be useful to keep the optional arguments?

Not of the top of my head, but if we have "Discard all", "Keep all", and "Keep `{}`" then also having "Keep `[]`" is trivial to add. So basicly it's a "why not" approach.

Ooh. Actually I do have a few environments where an optional argument is used as the title, so that is somewhere where I would want to use it.

> it seems that some replacements are done without being added to `this.changes`

Ah. I'll add the `\begin/end{document}` removals. Ooops. Think that's it though.
(done in 22022548101aa74702f70963954ca373ad79a035)

> You replace usually by nothing, which could lead grammar spellings to errors.

I'm afraid I didn't funnly think that through. Let's try that :) (`X` introduced in 1e5fe3ed8629009325da2ec3047da6b317851fcd)

>  replacement seems very similar to `this.change`, is it only for debugging?

Ahh, well I instroduced `replacements` just as a way to do the replacements all in one big batch to make some aspects easier. I didn't really think of the similarity till the end :stuck_out_tongue:
We could probably just switch out `replacements` for `this.changes`, we'd just need to also store the replacement text in `this.changes`.

> My point is that we had an "almost" working prototype and may be we should work iteratively from it, and not change everything in one go.

Yea, sorry to remove a large chunk of your work all of a sudden when it mostly worked :disappointed:.
Unfortunately, as I see it should be done for two reasons:
 - We need to match _balenced_ parenthesis, and unfortunately, all existing RegEx solutions use tokens such as `?R` which are unavailable in javascript's implementation :cry: (this can also be seen in you StackOverflow link)
 - 'Argument handling' is very nice, and the cleanest method I could think of required a new approach

> We could store a test file, its cleaned version and the translated positions for example.

Having something like that as a unit test sounds like a nice idea. I'm afraid I'm a bit lazy when it comes to setting things like that up, so for the moment I'll stick to my test file I mentioned earlier ([this one](https://github.com/tecosaur/BMC/blob/master/documentation/bmc.tex))


> There are issues with the given positions as input, so that the warning are not well translated:

> Since it worked before, I suspect the input. But also in the replace action, the cleaned version has some issues:

I think that may be better now.

> Are you sure you pushed everything? I tried your example and it also does not work. I tried with the old version and it mainly worked for your example, apart form some missing environments.

Yea.... I'm afraid it was just a bit half-baked. But! with you pointing out issues and me (hopefully) addressing them it should be much better :sweat_smile:

Thanks :)

**Comment by tecosaur** [comment]
Minor update, positions still don't seem to be right (off by two lines).
Not sure why though...
![image](https://user-images.githubusercontent.com/20903656/70367022-2cfc7800-18d7-11ea-92aa-4cb030cf7c07.png)

**Comment by tecosaur** [comment]
Update: everything seems to work (see below) if I remove a multi-line arg at the start

![image](https://user-images.githubusercontent.com/20903656/70367074-a5633900-18d7-11ea-849b-89249dceb35d.png)


![image](https://user-images.githubusercontent.com/20903656/70367059-8795d400-18d7-11ea-9629-6f94f2d1ee8b.png)

**Comment by tecosaur** [comment]
@PierreMarchand20 if you wouldn't mind, two things
 1. Could you have a look at how it is now and let me know how you find it?
 2. Might you have any idea what's going on with the issue I noticed in my last coments?

**Comment by PierreMarchand20** [comment]
In the link I sent you ([here](https://stackoverflow.com/a/35271017/5913047) more precisely) they show how to solve this issue in javascript.

I still have a lot of issues with my document. If I just look at the temporary file:
- `Let $\Omega \subset \mathbb{R}^d$ be a Lipschitz polyhedron, and \(d\) the geometric dimension.` becomes `Let $X X X^d$ be a Lipschitz polyhedron, and \ the geometric dimension`.
- magic comments are not deleted
- it looks like that instead of replacing math environments, it replaces what is inside

I mean, it is hard to say if there is an issue with the vscode ranges because I have environments that are half deleted, strange replacements by X, and commands that are not replaced: here's an example of temporary file I get
<img width="1008" alt="image" src="https://user-images.githubusercontent.com/15056323/70379694-b2a81400-1927-11ea-8c97-d0bfb7131536.png">

**Comment by PierreMarchand20** [comment]
I really think that it should be possible to do the modifications you want iteratively and starting with what I wrote.

**Comment by tecosaur** [comment]
@PierreMarchand20 hmm. That could be an idea, do you think we'd be able to achieve argument processing with that approach? I do think that going from the best-working solution and iterating on that sounds the best as long as we can get the same level of features (which is my concern).

Also re: regex, the only answers I found for javascript in that question were links to non-regex approaches ([here](https://stackoverflow.com/questions/15717436/js-regex-to-match-everything-inside-braces-including-nested-braces-i-want/27088184#27088184)) and suggestions to load user-written libraries.

#### _p.s._
_I recently handed in my glasses to get new lenses, so you probably won't see any new commits from me for the next few days; reading is hard, even at large font sizes (thanks astigmatism)  :cry:_

**Comment by PierreMarchand20** [comment]
Well my issue is that for now, the new version is far from working, and the approach is not that different from what I did. As far as I understood, you use a function to get matching parenthesis/brackets, but you could just use what I wrote and change the regex part by using this function without throwing everything away. Apart from that, you try to also parse the optional arguments and that's it no ? What are the additional features ?

Concerning regex: `XRegExp.matchRecursive` looked interesting, and we can still match a fix number of recursively (https://stackoverflow.com/a/17759264/5913047).

My point is that I cannot do much with the new version... I can try do add some new features to what I wrote because it is modular enough and it had a great differentiation of the types of commands. I can modify the part concerning commands whose arguments need to be processed, without breaking everything else.

**Comment by PierreMarchand20** [comment]
Good luck with the lenses then !

**Comment by tecosaur** [comment]
Yea, sorry about scrapping all of your hard work. I just thought at the time it would be easier. But then I also thought my code would 'just work', if only...

> What are the additional features?
 
The things I'd like to see are:
 - Different handling of different commands
 - Varying argument handling
 - Ditto for environments

> `XRegExp.matchRecursive`...

Perhaps, I'd generally am trying to minimise the number of libraries loaded (except for 'big' things like `csv-parser`). So I'll put this as a "maybe".

> My point is that I cannot do much with the new version...

Sorry about that

> I can try to add some new features to what I wrote because it is modular enough and it had a great differentiation of the types of commands. I can modify the part concerning commands whose arguments need to be processed, without breaking everything else.

Any improvements are most welcome as far as I'm concerned :)

Overall I'm in two minds about this (re: approach):
 1. Improve my revised version to the point that it _properly_ works, I feel as though it's close, but that could just be me
 2. My revision isn't working, we should just go back to your version and try to add the 'extras' that my version has. Still not sure as to how best deal with matching brackets, but I guess `XRegExp` looks like it should work.

> Good luck with the lenses then!

Thanks, hopefully, they get back to me sooner rather than later (I don't like realising how many hours I spend looking at text, and how bad my eyesight is :sweat_smile:)

**Comment by tecosaur** [comment]
I've been reunited with my glasses! So, I'm going to try to do some work on this over the next few days.
Have you got any thoughts about approach (1) vs (2)?

**Comment by PierreMarchand20** [comment]
It must be nice ^^

Well I don't know, it is up to you if you manage to fix your version. But I still think the version I did was good enough and working well. And now that I think about it, I am not even sure we need a recursive approach. For example: `\textbf{\textit{foo}}`, with my approach, I have a regex matching `\textbf{`, one matching `\textit{` and a regex matching the two `}`. With that, I have all the information I need I think. Since the text and the numbering of the characters are linear, it is enough.

**Comment by PierreMarchand20** [comment]
Because what is done afterward: I sort all the matching regex according to their start position. And for each warning coming from the linter, if the starting position of a regex is before, I translate the position of the warning by the length of the regex before.

**Comment by tecosaur** [comment]
Hmm. I might try looking more at what you had and fixing mine, then if that doesn't work out try vice versa and pulling features from my 'new' version back to yours.
I'll keep you updated. I may ping you to ask if/what issues you find at various points — how does that sound?

**Comment by PierreMarchand20** [comment]
Ok ok

**Comment by tecosaur** [comment]
... haven't forgotten, just got busy around Christmas.

**Comment by tecosaur** [comment]
And am now experiencing "stuff happens"+
Should get around to this eventually ... in the mean time I'll just release the current improvents in master.

**Comment by PierreMarchand20** [comment]
I am not sure it worth it to put it in the master for now, it is not really usable. Why don't you want to keep this branch ?

**Comment by tecosaur** [comment]
Oh, this is more an aside. I (think) I said earlier (somewhere) that I'd like to put this in the next release (or at least I was planning on doing so), so when I ambiguously said "current improvements" I was referring to the stuff _already in_ master (now in the v0.3.5 release). On second thoughts this could be quite confusing. Ooops.

Currently, I would like to see this mature, and then merge to master for a v0.4 release. However, unfortunately, lack the time to do so in the near future.
